### DISCUSSION TITLE
Custom Power Limits not working.

### DISCUSSION THREAD

**Discussion by paulusgi** [seerge/g-helper#226] [General]
Hi, I tested the 3 different power modes Silent, Balanced and Performance and the "Apply Power Limits" checkbox do nothing.

Whatever I set on the sliders the power consumption doesn't variate. Each power plan have their own limits but can't be changed.

I have an Idea about why is this happening but not sure: In armory crate when I want to modify PPT I have to set The "Manual" power profile. ¿Maybe with Silent, Balanced or Turbo my system don't allow the app the change the power limits?
I really don't know.

Tested with: Cinebench R23 and HWInfo64.
Laptop: Asus Rog Strix G513RM 2022 (Ryzen 7 6800H + RTX3060).
G-Helper Version 0.50
Under latest BIOS v327

**Comment by seerge** [comment]
@paulusgi can you post log from %appdata%/ghelper , it should mention if PPT settings are accepted by bios or not

**Reply by paulusgi** [reply]
Here it is :)   https://pastebin.com/b1T7JcD3

**Reply by seerge** [reply]
```
4/11/2023 10:42:22 PM: Boost 3
4/11/2023 10:42:23 PM: PowerLimit A0 = 125 : OK
4/11/2023 10:42:23 PM: PowerLimit A3 = 125 : OK
4/11/2023 10:42:23 PM: PowerLimit B0 = 80 : 0
```

I see that Platform limit (a0/a3) is accepted, and CPU limit - is not

It could be that on your laptop they are swapped (as it happened with some other models), and Platform (i.e. first slider) means CPU 
and second slider - does nothing

How do sliders look in armoury for you ? if it's only CPU related power limits - then it's probably the case

**Reply by paulusgi** [reply]
I don't have AC installed now to check. https://i.imgur.com/bMP6IEP.png

I will do some more testing because for some reasons sometimes the PPT works.

**Reply by seerge** [reply]
@paulusgi ok, seems like bios doesn't honour settings at all (despite answering 1 (OK) to settings)

There is a whole topic dedicated to endpoint discovery (i.e. how to set PPTs and other stuff) https://github.com/seerge/g-helper/issues/38 

You can also try to set limits by hand without any app too, in elevated powershell

```
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x001200A0, 25) 
(Get-WmiObject -Namespace Root/WMI -Class AsusAtkWmi_WMNB).DEVS(0x001200A3, 25) 
```

Where 25 would be your desired (CPU?) limit. App literally sets them in a same way (and i believe armoury too)

**Reply by paulusgi** [reply]
Thanks! The commands in elevated powershell doesn't work :S. So I ran the script and published my results on the topic #38 . <3

**Reply by seerge** [reply]
@paulusgi ok, so it's set somehow differently in your case then
you may try all 001200XX ids from my first post in #38 and try to set some limit there and see if any of them works (mb C1 C2 ?)

And i generally wonder if armoury custom limits work for you too :)

**Reply by paulusgi** [reply]
@seerge 
Hey I tried all of them wit no luck. Armory Crate in manual mode works for me and also there is another software that is working to limit PPT states.
That software:  https://github.com/JamesCJ60/Universal-x86-Tuning-Utility

Maybe this can help: https://github.com/JamesCJ60/Universal-x86-Tuning-Utility/blob/master/Scripts/Adpative%20Modes/Performance/CpuPowerLimiter.cs

**Reply by seerge** [reply]
@paulusgi that's strange, cause all endpoints exist on your laptop but by some reason are ignored
as for UXTU - i'm aware :) it's based on https://github.com/FlyGoat/RyzenAdj  as i understand and it works in a very different way comparing to g-helper / armoury + requires admin rights to run. And that's kind of a blocker. 

Armoury definitely works via asus driver / bios. Same as g-helper, too bad that not in your case :)

**Reply by seerge** [reply]
@paulusgi just to be sure, do you have latest version of Asus System control interface ?

**Reply by paulusgi** [reply]
Yes I have it.
Version V3.1.12.0

I edited my post on #38 adding my dsdt file. @seerge  Maybe it can help.

**Reply by paulusgi** [reply]
@seerge Hi! I tested Armory Crate manual mode and it is working! The sliders change the STAPM_Limit PPT_Limit_Fast and PPT_Limit_Slow.
Here how the sliders looks:
![acsliderscpu](https://user-images.githubusercontent.com/102924649/231720427-f9fce9a7-51e6-42f2-9563-8473cb487156.png)

I ran the script again just in case... https://pastebin.com/hAYWwXQ3

So maybe my laptop uses a different way/endpoint to apply the ppt? So Are there any other ways to find endpoints? Thanks you so much.

Edit: I found the "Set PPT Limit" inside the ssdt10.dsl limit: https://pastebin.com/pHa6Z4Pc  maybe you can find somthing interesting in there.

**Reply by paulusgi** [reply]
Hellow @seerge 
I finally figured out how to fix it. The truth has been nonsense... But if I only activate "Apply Power Limits" the PPT values ​​are ignored BUT.... if I activate "Apply custom curved fan" and "Apply Power Limits" at the same time, then the ppt values ​​are applied and working. So I must use a custom fan curve if I want to use a custom power limits.

**Reply by seerge** [reply]
@paulusgi , ok, that's surprising .... but explains why it works in armoury (there is no way to set them separately.

TUF laptops on intel CPU had opposite issue, setting FANs were breaking PPTs (if they were not applied).

I can hardcode that for your model (models?) app will force custom fan curves ? will that be a solutuion?

**Reply by seerge** [reply]
@paulusgi do custom fan curves without custom PPTs work on your laptop?

**Reply by paulusgi** [reply]
> @paulusgi , ok, that's surprising .... but explains why it works in armoury (there is no way to set them separately.
> 
> TUF laptops on intel CPU had opposite issue, setting FANs were breaking PPTs (if they were not applied).
> 
> I can hardcode that for your model (models?) app will force custom fan curves ? will that be a solution?

@seerge 
Yes it will be a solution, BUT the order is important. I mean if I activate custom power limits first and then I activate custom fan curves nothing happens and the ppt values are not set. So I must apply the custom fan curves first and then apply the custom power limits to make them properly work.

And about the fan curves without PPTs yes they work without the PPTs values.

**Reply by paulusgi** [reply]
In short:
- Apply custom fan curves ONLY (NO ppts checkbox marked) = Custom fan curve works
- 1.Apply custom fan curves -> 2. Apply custom PPTs = Both will work if applied in that order (1->2).
- Apply custom PPTs without fan curves or apply custom curves after applying ppts = ONLY custom fan curves will work (PPTs will NOT)

**Reply by seerge** [reply]
[GHelper.zip](https://github.com/seerge/g-helper/files/11224793/GHelper.zip)
Can you try this build ? app will run fan-curve command before PPTs (unless "real" custom fan curve is applied)

**Comment by paulusgi** [comment]
@seerge okey did it. The app performs exactly as the 0.51. When I only apply the custom power profile nothing happens. The custom fan curve are not activated when "Apply Power Limits" are selected. :S So for me is exactly the same app.
EDIT
SORRY, after 5 minuts the PPTs values as been applied without the "Apply Custom Fan Curve" checked.

**Reply by seerge** [reply]
@paulusgi may be it needs some delay before setting PPTs after fans - try this one
[GHelper.zip](https://github.com/seerge/g-helper/files/11225221/GHelper.zip)

**Reply by paulusgi** [reply]
> @paulusgi may be it needs some delay before setting PPTs after fans - try this one [GHelper.zip](https://github.com/seerge/g-helper/files/11225221/GHelper.zip)

@seerge 


So if the fan curve its not modified (not need to be applied) Apply Custom PPTs doesn't work.

**Reply by seerge** [reply]
@paulusgi clear, thanks - try this one, it should work even when you delete config
[GHelper.zip](https://github.com/seerge/g-helper/files/11225505/GHelper.zip)

(problem was in applying a fan curve that didn't exist before you edited it at least somehow, now it will save it as soon as it opened for the first time)

**Reply by paulusgi** [reply]
> @paulusgi clear, thanks - try this one, it should work even when you delete config [GHelper.zip](https://github.com/seerge/g-helper/files/11225505/GHelper.zip)
> 
> (problem was in applying a fan curve that didn't exist before you edited it at least somehow, now it will save it as soon as it opened for the first time)

@seerge YES! It works perfectly without having to modify the graphical fan curve or selecting the "Apply Custom Fan Curve" checkbox.
Now the software will properly apply the PPTs on laptops like mine or similar :D
Thank you so much for your patience.

**Reply by seerge** [reply]
@paulusgi great, the only problem is that Platform PPT probably affects CPU limit and CPU slider does nothing ?

**Reply by paulusgi** [reply]
> @paulusgi great, the only problem is that Platform PPT probably affects CPU limit and CPU slider does nothing ?

@seerge 
Exactly the CPU slieder does nothing. And the Platform slieder affects 2 things: the STAPM_LIMIT and the PPT_LIMIT_SLOW.

**Reply by seerge** [reply]
@paulusgi that's expected based on weird DSL implementation in your case

Current app logic is to check if B0 endpoint exists (it's a second slider) and if so - show it
Otherwise show only one A0 slider (if it also exists)

Problem that your BIOS reports both A0 and B0 to exist ... :) the only fix would be hardcoding your model number again (which i'm not a big fan of), so I will probably keep it as is for now.

I have pushed fixes as part of 0.52 https://github.com/seerge/g-helper/releases/tag/v0.52 and closing this as "solved" for now.

Thanks for help!

**Reply by paulusgi** [reply]
Its more than enough so... big thanks again now it works perfect. <3 ^_^
